### PR TITLE
drivers: udc_stm32: Implement driver API to get actual device speed and fix ZLP handling

### LINKED ISSUE #68817: usb: udp: frames of exact multiples of 64 Byte not transmitted correctly
Hi there,

**Describe the bug**

If a UDP package is sent out using `zsock_send` and the length of the frame on the wire is a multiple of 64 Byte (128, 192, or 256, and so on) the package is not transmitted. Instead the package seems to be buffered. Next calls of `zsock_send` seem also to be buffered, until the sum of all the buffered data is no multiple of 64 Bytes. Then the first package is sent out and all the other data appears in the ethernet trailer. At least it looks like that in Wireshark.

**To Reproduce**

Sniff with wireshark on USB. Do something like the following code on the MCU.

```c
static char my_63_data[21];
static char my_64_data[22/*+64*/];
static struct sockaddr dst_addr;
static int sock_fd;

void some_init()
{
	/* set some dst_addr */
	/*...*/

	zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
	zsock_connect(sock_fd, dst_addr, sizeof(struct sockaddr_in));
	int flags = zsock_fcntl(sock_fd, F_GETFL, 0);
	zsock_fcntl(sock_fd, F_SETFL, flags | O_NONBLOCK);

	snprintf(&my_63_data, sizeof(my_63_data), "63B 63B 63B 63B 63B ");
	snprintf(&my_64_data, sizeof(my_64_data), "64B 64B 64B 64B 64B 4");
}

void send_64_byte(void)
{
	zsock_send(sock_fd, &my_64_data, sizeof(my_64_data), 0);
}

void send_63_byte(void)
{
	zsock_send(sock_fd, &my_63_data, sizeof(my_63_data), 0);
}
```

Call the functions `send_63_byte` and `send_64_byte` several times in some random order.

**Expected behavior**
For each zsock_send call a frame in wireshark.

**Impact**
Some packages get dropped, if the application waits for a response it will timeout eventually or wait forever.

**Logs and console output**
grafik
The "+64" in the example has been uncommented, that's why the payload longer and filled with more zeros. Then `send_63_byte`, `send_64_byte`, and `send_63_byte` have been called in this order.

**Environment**

I use Zephyr OS @ a8a0eb52439caccf51c66f845b2c29ceef0da39e on an STM32H730VB MCU with a gcc-arm-none-eabi-10.3-2021.07 toolchain.

Relevant config:
```
CONFIG_NETWORKING=y
CONFIG_NET_IPV6=n
CONFIG_NET_IPV4=y
CONFIG_NET_UDP=y

CONFIG_USB_DEVICE_STACK_NEXT=y
CONFIG_NET_L2_ETHERNET=y
CONFIG_USBD_CDC_ECM_CLASS=y

CONFIG_NET_SOCKETS=y
```

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by jfischer-no**
Implement driver API to get actual device speed and handle ZLP flag.
Both of these fixes are required for the driver to work properly.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/68817

### CODE REVIEW

**Review by erwango — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by tmon-nordic — APPROVED** [review]

**Comment by GeorgeCGV** [comment]
Fixes #68817.

File: drivers/usb/udc/udc_stm32.c [hunk 1]
```diff
@@ -768,6 +775,21 @@ static int udc_stm32_ep_dequeue(const struct device *dev,
 	return 0;
 }
 
+static enum udc_bus_speed udc_stm32_device_speed(const struct device *dev)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+
+	if (priv->pcd.Init.speed == USBD_HS_SPEED) {
+		return UDC_BUS_SPEED_HS;
```
Comment (fabiobaltieri) [inline drivers/usb/udc/udc_stm32.c:783]: looks like these constant are not defined in the HAL for F1 series mcus, should this go under an ifdef?

Comment (henrikbrixandersen) [reply drivers/usb/udc/udc_stm32.c:783]: They are defined in `include/zephyr/drivers/usb/udc.h`?

Comment (fabiobaltieri) [reply drivers/usb/udc/udc_stm32.c:783]: sorry my bad, wrong line, yes

Comment (fabiobaltieri) [reply drivers/usb/udc/udc_stm32.c:783]: Opened https://github.com/zephyrproject-rtos/zephyr/pull/75306

Comment (jfischer-no) [reply drivers/usb/udc/udc_stm32.c:783]: > looks like these constant are not defined in the HAL for F1 series mcus, should this go under an ifdef?

I did not expect that, they define USBD_FS_SPEED for F1 but not USBD_HS_SPEED. @fabiobaltieri do you have a PR ready or should I fix it?

Comment (fabiobaltieri) [reply drivers/usb/udc/udc_stm32.c:783]: yeah, guess the new stack needs a bit more CI build coverage, those HALs are evil they have all sort of inconsistencies

**Pull request merged**